### PR TITLE
Create kubevirt_vmi_launcher_memory_overhead_bytes metric

### PR DESCRIPTION
**Pull request by machadovilaca**
### What this PR does
Before this PR:

No visibility over calculate overhead memory for vmi launchers

After this PR:

Metric with  overhead memory for vmi launchers value

jira-ticket: https://issues.redhat.com/browse/CNV-41807

Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
Create kubevirt_vmi_launcher_memory_overhead_bytes metric
```

### CODE REVIEW

**Comment by machadovilaca** [comment]
@iholder101 @maiqueb can you confirm that the result of `services.GetMemoryOverhead` is the value we are looking for?

**Comment by iholder101** [comment]
> @iholder101 @maiqueb can you confirm that the result of `services.GetMemoryOverhead` is the value we are looking for?

Generally yes, but bear in mind that there are exceptions, for example if the VM has use pages:
https://github.com/kubevirt/kubevirt/blob/55fc278af78098e9f9ebb2c80f57209114b838ac/pkg/virt-controller/services/template.go#L1443-L1468

File: pkg/virt-controller/services/template.go [hunk 1]
```diff
@@ -1447,6 +1448,7 @@ func (t *templateService) VMIResourcePredicates(vmi *v1.VirtualMachineInstance,
 		vmiCPUArch = t.clusterConfig.GetClusterCPUArch()
 	}
 	memoryOverhead := GetMemoryOverhead(vmi, vmiCPUArch, t.clusterConfig.GetConfig().AdditionalGuestMemoryOverheadRatio)
+	metrics.SetVmiLaucherMemoryOverhead(vmi, memoryOverhead)
```
Comment (iholder101) [inline pkg/virt-controller/services/template.go:1451]: Aren't we supposed to keep the metric bits under `./pkg/monitoring`?

Comment (machadovilaca) [reply pkg/virt-controller/services/template.go:1451]: At this stage, only metrics definitions and more Prometheus-related concepts, for example, in this example we are not using labels nor setting the values directly. Everything is handled by the `SetVmiLaucherMemoryOverhead` function.

The alternative here would be a collector to fetch the virt-launcher requested memory minus the domain requested memory. Would that be accurate?

Comment (iholder101) [reply pkg/virt-controller/services/template.go:1451]: > The alternative here would be a collector to fetch the virt-launcher requested memory minus the domain requested memory. Would that be accurate?

Unfortunately it wouldn't. For example, if both memory request and guest memory are explicitly set (e.g. request=2Gi, guest=2Gi. the pod is overcommitted, but you can't know the calculated infra overhead).

I tend to think that the guest's calculated overhead is important enough in order to expose it. We can do so either with a new status field, or even with an annotation. I think that by doing so we'd gain:
* Eliminate the addition of monitoring code into non `./pkg/monitoring` packages.
* Let users understand the full picture w.r.t. the VM's memory calculations by looking at the VM's manifest only (i.e. not depending on Prometheus metrics for that):
  * The requested memory. (exists today)
  * The guest memory. (exists today)
  * The calculated infra overhead.
  * (And hence ->) the overcommitment ratio.
    * (should we expose that as well via status/annotation?)

@fabiand @vladikr Do you have any thoughts regarding this?

Comment (machadovilaca) [reply pkg/virt-controller/services/template.go:1451]: maybe, not to block following work depending on this metric, we can move forward with this implementation, and I can open a follow-up issue/PR to discuss the alternatives

Comment (iholder101) [reply pkg/virt-controller/services/template.go:1451]: Sure, sounds good :+1: 
please cc me in that issue :)

File: docs/metrics.md [hunk 2]
```diff
@@ -120,6 +120,9 @@ Information about VirtualMachineInstances. Type: Gauge.
 ### kubevirt_vmi_last_api_connection_timestamp_seconds
 Virtual Machine Instance last API connection timestamp. Including VNC, console, portforward, SSH and usbredir connections. Type: Gauge.
 
+### kubevirt_vmi_launcher_memory_overhead_bytes
+Estimation of additional memory (over the requested) needed for the domain to operate properly. Type: Gauge.
```
Comment (iholder101) [inline docs/metrics.md:124]:
```suggestion
Estimation of the memory amount required for virt-launcher's infrastructure components (e.g. libvirt, QEMU). Type: Gauge.
```

**Review by iholder101 — COMMENTED** [review]
Thank you @machadovilaca! :)
Left a minor comment.

Please let someone else from sig-observability to review the monitoring bits.
/approve

File: docs/metrics.md (hunk 2 above)
Comment (machadovilaca) [reply docs/metrics.md]: updated

**Comment by pbxqdown** [comment]
Hey folks, I'm wondering whether kubevirt_vmi_launcher_memory_overhead_bytes contains metrics for deleted vmi? We found in our environment kubevirt_vmi_launcher_memory_overhead_bytes metrics count is big, it seems includes metrics for deleted vmis.

**Pull request merged**